File: src/main/resources/Resource/GreedTest.cpp
${<if HasArray}
// Array to vector converter
template <typename T> vector<T> vector_wrapper(T arr[], int n) { return vector<T>(arr, arr + n); }
#define to_vector(arr) vector_wrapper(arr, sizeof(arr) / sizeof(arr[0]))
${<end}

template <typename T> string pretty_print(T t) { stringstream s; typeid(T) == typeid(string) ? s << "\\"" << t << "\\"" : s << t; return s.str(); }

${<if ReturnsArray}
// Vector print
template <typename T> ostream &operator << (ostream &out, vector<T> arr) {
    out << "{ ";
    for (int i = 0; i < arr.size(); ++i) out << (i == 0 ? "" : ", ") << pretty_print(arr[i]);
    out << " }";
    return out;
}
${<end}

int nPassed = 0, nAll = 0;
int nExample = ${NumOfExamples}, nCustom = 0;

void do_test(${Method.Params}, ${Method.ReturnType} expected, int caseNo = -1) {
    nAll++;
    bool isExample = caseNo >= 0 && caseNo < nExample;
    cout << (isExample ? "    Custom " : "    Example") << " #" << (isExample ? caseNo : nCustom++) << " ... ";

${<if RecordRuntime}
    time_t startClock = clock();
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
${<end}
    ${ClassName} instance;
    ${Method.ReturnType} result = instance.${Method.Name}(${foreach Method.Params par , }${par.Name}${end});
${<if RecordRuntime}
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
${<end}

    if (result == expected) {
        nPassed++;
        cout << "PASSED!" ${if RecordRuntime}<< " (" << elapsed << " seconds)" ${end}<< endl;
    }
    else {
        cout << "FAILED!" ${if RecordRuntime}<< " (" << elapsed << " seconds)" ${end}<< endl;
        cout << "            Expected: " << pretty_print(expected) << endl;
        cout << "            Received: " << pretty_print(result) << endl;
    }
}

void run_testcase(int no) {
    switch (no) {
${<foreach Examples e}
        case ${e.Num}: {
${<foreach e.Input in}
${<if !in.Param.Type.Array}
            ${in.Param.Type.Primitive} ${in.Param.Name} = ${in};
${<else}
            ${in.Param.Type.Primitive} ${in.Param.Name}[] = {${foreach in.ValueList v ,}
                ${v}${end}
            };
${<end}
${<end}
${<if !e.Output.Param.Type.Array}
            ${e.Output.Param.Type.Primitive} expected = ${e.Output};
${<else}
            ${e.Output.Param.Type.Primitive} expected[] = {${foreach e.Output.ValueList v ,}
                ${v}${end}
            };
${<end}
            do_test(${foreach e.Input in , }${if in.Param.Type.Array}to_vector(${in.Param.Name})${else}${in.Param.Name}${end}${end}, ${if e.Output.Param.Type.Array}to_vector(expected)${else}expected${end}, no);
            break;
        }
${<end}

        // Your custom testcase goes here
        case -1:
            break;
        default: break;
    }
}

int main(int argc, char *argv[]) {
    cout << "${Problem.Name} (${Problem.Score} Points)" << endl << endl;
    if (argc == 1)
        for (int i = 0; i < nExample; ++i) run_testcase(i);
    else
        for (int i = 1; i < argc; ++i) run_testcase(atoi(argv[i]));

    cout << endl << "Passed " << nPassed << "/" << nAll << endl;
    return 0;
}

File: src/main/resources/Resource/GreedTest.java
	static void runTestcase(int cs) {
		switch (cs) {
			// Your custom testcase goes here
			case -1:
				break;

${<foreach Examples e}
			case ${e.Num}: {
${<foreach e.Input in}
${<if !in.Param.Type.Array}
				${in.Param.Type.Primitive} ${in.Param.Name} = ${in};
${<else}
				${in.Param.Type.Primitive} ${in.Param.Name}[] = new ${in.Param.Type} {${foreach in.ValueList v ,}
					${v}${end}
				};
${<end}
${<end}
${<if !e.Output.Param.Type.Array}
				${e.Output.Param.Type.Primitive} expected = ${e.Output};
${<else}
				${e.Output.Param.Type.Primitive} expected[] = new ${e.Output.Param.Type} {${foreach e.Output.ValueList v ,}
					${v}${end}
				};
${<end}
				doTest(${foreach e.Input in , }${in.Param.Name}${end}, expected, cs);
				break;
			}
${<end}
			default: break;
		}
	}

	static void doTest(${Method.Params}, ${Method.ReturnType} expected, int caseNo) {
${<if RecordRuntime}
		long startTime = System.currentTimeMillis();
${<end}
		Throwable exception = null;
		${ClassName} instance = new ${ClassName}();
		${Method.ReturnType} result = ${Method.ReturnType;ZeroValue};
		try {
			result = instance.${Method.Name}(${foreach Method.Params par , }${par.Name}${end});
		}
		catch (Throwable e) { exception = e; }
${<if RecordRuntime}
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;
${<end}

		nAll++;
		boolean isExample = caseNo >= 0 && caseNo < nExample;
		if (isExample)
			System.err.print(String.format("   Example #%d ...", caseNo));
		else
			System.err.print(String.format("   Custom  #%d ...", nCustom++));

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
		}
		else if (${if Method.ReturnType.Array}equals(result, expected)${else}${if Method.ReturnType.String}expected.equals(result)${else}${if Method.ReturnType.RealNumber}Math.abs(result - expected) < 1e-9${else}result == expected${end}${end}${end}) {
			System.err.println("PASSED! "${if RecordTime} + String.format("(%.2f seconds)", elapsed)${end});
			nPassed++;
		}
		else {
			System.err.println("FAILED! "${if RecordTime} + String.format("(%.2f seconds)", elapsed)${end});
			System.err.println("            Expected: " + ${if Method.ReturnType.Array}toString(expected)${else}expected${end});
			System.err.println("            Received: " + ${if Method.ReturnType.Array}toString(result)${else}result${end});
		}
	}

	static int nExample = ${NumOfExamples}, nCustom = 0;
	static int nAll = 0, nPassed = 0;

${<if ReturnsArray}
	static boolean equals(${Method.ReturnType} a, ${Method.ReturnType} b) {
		if (a.length != b.length) return false;
		for (int i = 0; i < a.length; ++i) if (${if Method.ReturnType.String}a[i] == null || b[i] == null || !a[i].equals(b[i])${else}${if Method.ReturnType.RealNumber}Math.abs(a[i] - b[i]) > 1e-9${else}a[i] != b[i]${end}${end}) return false;
		return true;
	}

	static String toString(${Method.ReturnType} arr) {
		StringBuffer sb = new StringBuffer();
		sb.append("[ ");
		for (int i = 0; i < arr.length; ++i) {
			if (i > 0) sb.append(", ");
			sb.append(arr[i]);
		}
		return sb.toString() + " ]";
	}
${<end}
	public static void main(String[] args){
		System.err.println("${Problem.Name} (${Problem.Score} Points)");
		System.err.println();
		if (args.length == 0)
			for (int i = 0; i < nExample; ++i) runTestcase(i);
		else
			for (int i = 0; i < args.length; ++i) runTestcase(Integer.parseInt(args[i]));
		System.err.println(String.format("%nPassed %d/%d", nPassed, nAll));
	}
